### ISSUE TITLE
Report inference failures as errors by default

### ISSUE THREAD

**Issue by msridhar** [uber/NullAway#1551]
After #1549, there are cases where inference fails but the default type inferred by `javac` does not lead to any NullAway errors being reported, for the case of captures, etc.  See the test `com.uber.nullaway.jspecify.WildcardTests#superWildcardToConcreteTypeVariable`.

This exposes a deeper issue with our error reporting, which is that we do not report inference failures by default as an error.  In most cases, we get away with this, as some other error gets reported by NullAway given the default `@NonNull` nullability for the type variables.  But, this error message is a bit misleading (since we did not successfully infer `@NonNull`), and, as shown by the case above, it's a hack that doesn't always work.

The right solution here is to report inference failures as errors by default and to suppress warnings that are subsumed by the inference failure as best we can.  We also need to improve our inference failure warning messages as right now they completely lack information.

**Comment by msridhar** [comment]
**This comment is outdated and incorrect**.  We had an incorrect idea of how to fix this issue before, as seen below.  Keeping it for the record.

We should explicitly create a substitution and map all type variables to `@NonNull` to ensure errors get reported for such cases.

Plan from Codex:

```
  1. Introduce a fallback substitution for InferenceFailure.

     Instead of returning javac’s call-site method type unchanged in substituteTypeArgsInGenericMethodType(...), build a conservative NullAway fallback where
     each method type variable is treated as @NonNull.
  2. Use that fallback only for downstream checking, not for reporting inference success.

     Keep InferenceFailure as a failure result, but when later code needs a method type for parameter compatibility or return/assignment checks, substitute
     failed method type vars with synthetic @NonNull type vars or otherwise annotate the call-site substitutions as non-null.
  3. Be careful with nullable upper bounds.

     Even if a method type variable has T extends @Nullable Object, the fallback should still probably choose @NonNull T for unknown inference, matching the
     existing successful-inference behavior for unconstrained variables. But this should be explicitly documented because it is a policy choice.
  4. Avoid duplicate diagnostics when warnings are enabled.

     If warnOnGenericInferenceFailure is true, we probably do not want both:
      - Failed to infer type argument nullability
      - plus a downstream parameter/assignment mismatch caused by fallback defaults

     We can either suppress downstream fallback diagnostics when the inference warning is emitted, or decide the concrete mismatch is better and stop reporting
     the generic inference warning for that call.
  5. Add tests for both modes.

     For the new wildcard test:
      - With inference-failure warnings enabled, assert the current inference-failure diagnostic or the chosen replacement diagnostic.
      - With warnings disabled, assert a concrete parameter/assignment nullability error still appears.
  6. Keep fallback local to failed inference.

     Don’t change the successful inference path. The existing putIfAbsent(..., NONNULL) default for unconstrained variables should remain as-is.
```
